dali/kernels/erase/erase_gpu.h

```diff
@@ -0,0 +1,526 @@
+  }
+};
+
+struct do_erase {
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, span<T> fill_values) {
+    (void)erase_regions;
+    if (channel_dim == -1) {
+      sample.out[offset] = fill_values[0];
+    } else {
+      sample.out[offset] = fill_values[coordinate[channel_dim]];
+    }
+  }
+};
+
+struct do_copy_or_erase {
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, span<T> fill_values) {
+    auto fill_value = channel_dim == -1 ? fill_values[0] : fill_values[coordinate[channel_dim]];
+    copy_or_erase<channel_dim>(sample, erase_regions, coordinate, offset, fill_value);
+  }
+
+  template <int channel_dim, typename T, int ndim>
+  __device__ static void copy_or_erase(erase_sample_desc<T, ndim> sample,
+                                       span<ibox<ndim>> erase_regions, ivec<ndim> coordinate,
+                                       int64_t offset, T fill_value) {
+    bool copy = true;
+    for (auto &region : erase_regions) {
+      if (region.contains(coordinate)) {
+        copy = false;
+      }
+    }
+    if (copy) {
+      sample.out[offset] = sample.in[offset];
+    } else {
+      sample.out[offset] = fill_value;
+    }
+  }
+};
+
+template <int channel_dim = -1, typename T, int ndim = 2>
+__global__ void erase_gpu_impl(erase_sample_desc<T, ndim> *samples, ivec<ndim> region_dims,
+                               int erase_regions_count, span<T> fill_values) {
+  const auto region_idx = blockIdx.x;
+  const auto sample_idx = blockIdx.y;
+
+  const auto &sample = samples[sample_idx];
+  ivec<ndim> sample_dims = sample.sample_shape;
+
+  auto region_box = get_region(region_idx, region_dims, sample_dims);
+
+  const auto *erase_regions_ptrs = sample.erase_regions;
+  constexpr int max_regions = 1024;
+  __shared__ ibox<ndim> erase_regions[max_regions];
+  __shared__ int filtered_region_idx;
+
+  filtered_region_idx = 0;
+  __syncthreads();
+  int full_cover = false;
+
+  // check which erase_regions overlap with current region_box
+  for (int i = threadIdx.y * blockDim.x + threadIdx.x; i < erase_regions_count;
+       i += blockDim.y * blockDim.x) {
+    if (erase_regions_ptrs[i].overlaps(region_box)) {
+      erase_regions[atomicAdd(&filtered_region_idx, 1)] = erase_regions_ptrs[i];
+      if (erase_regions_ptrs[i].contains(region_box)) {
+        full_cover = true;
+        break;
+      }
+    }
+  }
+  __syncthreads();
+
+  //  Check if we have full or no cover at all
+  int total_erase_regions = filtered_region_idx;
+  __syncthreads();
```

Comment:
Why we need this additional sync here?

Reply:
After accessing the shared memory, but it's probably not needed here.
@mzient?

Reply:
It is read so I think ti should be safe.

Reply:
Removed